### PR TITLE
[6.0] New Numeric methods

### LINKED ISSUE #486: Add Numeric toBytes, toBits
**SNumeric**: 
- `toBytes`; (critical)
- `toBits`; (critical)

### LINKED ISSUE #1006: New Numeric methods for 6.0
From #479 :
* abs 
* compare

From 
*`~` (bitwise inversion);
* `|` (bitwise OR for Numeric);
* `&` (bitwise AND for Numeric);
* `^` (bitwise XOR for Numeric);
* `<<`, `>>`, `>>>` (bit shifts for Numeric);

### PR DESCRIPTION
**Pull request by kushti**
In this PR, new numeric methods are implemented for 6.0:

* .toBigEndianBytes
* .toBits
* .bitwiseInverse
* .bitwiseOr
* .bitwiseAnd
* .bitwiseXor
* .shiftLeft
* .shiftRight

close #486 
close #1006

### CODE REVIEW

File: data/shared/src/main/scala/sigma/data/ExactNumeric.scala [hunk 1]
```diff
@@ -30,6 +32,66 @@ trait ExactNumeric[T] {
...
+  def toBigEndianBytes(x: T): Coll[Byte]
+
+  /**
+    * Returns a big-endian binary representation of this value as boolean array.
+    */
+  def toBits(x: T): Coll[Boolean] = {
+
+    def isBitSet(byte: Byte)(bit: Int): Boolean = ((byte >> bit) & 1) == 1
+
+    def byte2Bools(b: Byte): Array[Boolean] =  (0 to 7).toArray.reverse.map(isBitSet(b))
```
Comment (aslesarenko) [inline data/shared/src/main/scala/sigma/data/ExactNumeric.scala:48]: this methods creates 1 Range allocation + 3 allocations of arrays.

File: data/shared/src/main/scala/sigma/data/ExactNumeric.scala [hunk 2]
```diff
@@ -30,6 +32,66 @@ trait ExactNumeric[T] {
...
+
+    val bytes = toBigEndianBytes(x)
+    val l = bytes.length
+    val res = new Array[Boolean](l * 8)
+    var offset = 0
+    cfor(0)(_ < l, _ + 1) { i =>
+      val b = bytes(i)
+      val bits = byte2Bools(b)
```
Comment (aslesarenko) [inline data/shared/src/main/scala/sigma/data/ExactNumeric.scala:56]: It is possible to allocate `bits` once and then reuse as a buffer. 
Or even better, put boolean values directly into `res`.
It can be filled from inner cfor loop over bit indexes `cfor(7, _ >= 0, _ - 1)`.

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala [hunk 3]
```diff
@@ -48,129 +51,787 @@ class LanguageSpecificationV6 extends LanguageSpecificationBase { suite =>
...
+          SByteMethods.v6Methods.find(_.name == "shiftLeft").get,
+          Vector(SelectField.typed[Value[SInt.type]](ValUse(1, SPair(SByte, SInt)), 2.toByte)),
+          Map()
+        )
+      ),
+      sinceVersion = V6SoftForkVersion)
+
+    verifyCases(
+      Seq(
+        (3.toByte, 3) -> new Expected(ExpectedResult(Success(24.toByte), None))
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:206]: I suggest to add more cases, in particular with boundary conditions and negative values. Also to cover error cases.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: data/shared/src/main/scala/sigma/ast/methods.scala [hunk 4]
```diff
@@ -244,7 +387,22 @@ object SNumericTypeMethods extends MethodsContainer {
...
+    BitwiseOrMethod,
+    BitwiseAndMethod,
+    BitwiseXorMethod,
+    ShiftLeftMethod,
+    ShiftRightMethod
+  )
+
+  protected override def getMethods(): Seq[SMethod] = {
+    // this .getMethods shouldn't ever be called
+    ???
```
[the remark this answers is not in the document]
Comment (kushti) [reply data/shared/src/main/scala/sigma/ast/methods.scala:401]: done

File: data/shared/src/main/scala/sigma/data/ExactNumeric.scala (hunk 1 above)
Comment (kushti) [reply data/shared/src/main/scala/sigma/data/ExactNumeric.scala:48]: optimized

File: data/shared/src/main/scala/sigma/data/ExactNumeric.scala (hunk 2 above)
Comment (kushti) [reply data/shared/src/main/scala/sigma/data/ExactNumeric.scala:56]: done

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala (hunk 3 above)
Comment (kushti) [reply sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:206]: done

**Comment by kushti** [comment]
@aslesarenko comments addressed, please make another pass

**Review by aslesarenko — APPROVED** [review]
